src/misc/umap.c

```diff
@@ -0,0 +1,262 @@
+            IGRAPH_CHECK(igraph_umap_decay(&weight_b, distance, VECTOR(*open_set_sizes)[j], VECTOR(*open_set_decays)[j]));
+            if (weight_a <= 0 || weight_b <= 0)
+                continue;
+            weight = weight_a + weight_b - weight_a * weight_b;
+            IGRAPH_CHECK(igraph_vector_int_push_back(&edges, i));
+            IGRAPH_CHECK(igraph_vector_int_push_back(&edges, j));
+            IGRAPH_CHECK(igraph_vector_push_back(umap_weights, weight));
+        }
+    }
+
+    IGRAPH_CHECK(igraph_create(umap_graph, &edges, no_of_nodes, IGRAPH_UNDIRECTED));
+    igraph_vector_int_destroy(&edges);
+    IGRAPH_FINALLY_CLEAN(1);
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_get_gradient(igraph_matrix_t *gradient, igraph_matrix_t *layout, igraph_t *umap_graph, igraph_vector_t *umap_weights)
+{
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(layout);
+    igraph_vector_int_t eids;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&eids, 0);
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        igraph_incident(umap_graph, &eids, i, IGRAPH_ALL);
+        MATRIX(*gradient, i, 0) = 0;
+        MATRIX(*gradient, i, 1) = 0;
+        for (igraph_integer_t j = 1; j < igraph_vector_int_size(&eids); j++) {
+            igraph_integer_t eid = VECTOR(eids)[j];
+            igraph_real_t x = MATRIX(*layout, i, 0);
+            igraph_real_t y = MATRIX(*layout, i, 1);
+            igraph_integer_t other = IGRAPH_OTHER(umap_graph, eid, i);
+            igraph_real_t other_x = MATRIX(*layout, other, 0) ;
+            igraph_real_t other_y = MATRIX(*layout, other, 1) ;
+            igraph_real_t x_diff = (x - other_x);
+            igraph_real_t y_diff = (y - other_y);
+            igraph_real_t distance = (x_diff * x_diff + y_diff * y_diff);
+            igraph_real_t weight_diff = -VECTOR(*umap_weights)[j] - 1 / distance;
+            igraph_real_t gradient_x = weight_diff * x_diff / distance;
+            igraph_real_t gradient_y = weight_diff * y_diff / distance;
+            MATRIX(*gradient, i, 0) += gradient_x;
+            MATRIX(*gradient, i, 1) += gradient_y;
+        }
+    }
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_umap_layout(igraph_t *umap_graph, igraph_vector_t *umap_weights,
+        igraph_matrix_t *layout) {
+    igraph_integer_t epochs = 100;
+    igraph_real_t learning_rate = 0.1;
+    igraph_matrix_t gradient;
+    igraph_matrix_init(&gradient, igraph_matrix_nrow(layout), igraph_matrix_ncol(layout));
+
+    igraph_layout_random(umap_graph, layout);
+    for (igraph_integer_t e = 0; e < epochs; e++) {
+        igraph_get_gradient(&gradient, layout, umap_graph, umap_weights);
+        igraph_matrix_scale(&gradient, learning_rate);
+        igraph_matrix_sub(layout, &gradient);
+    }
+
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return (IGRAPH_SUCCESS);
+}
+
+igraph_error_t igraph_umap(igraph_matrix_t *data, igraph_matrix_t *layout) {
+    igraph_t knn_graph;
```

Comment:
As above, ideally we start directly from a similarity graph (call it knn if you like the name). How that is constructed should not be our business I believe?

Reply:
(I'm using this process: https://umap-learn.readthedocs.io/en/latest/how_umap_works.html)
I used a graph twice, first to get a KNN graph, and then I used that graph to create a new graph, which is then used in the layout process. Do you mean the function should accept the first graph, which should then be transformed using the fuzzy open sets, which should then generate the new graph?
Or should this function only do the minimization of the cross entropy?

Reply:
Good point, I'm referring to the former here: we do not need to build the knn graph from a vector space in igraph I think (but am open to other opinions). As for option 2, even though having a function that does purely operation 2 might be useful in the future, ATM it sounds too specialized.

Since we are here, one word on the article. Basically our starting point - the way the current implementation works as well -  is a weighted **truncated** knn graph, i.e. a similarity graph in which each vertex/cell is connected (with weight equal to some measure, i.e. correlation) to its k neighbors *unless you run out of neighbors above a certain similarity* (tipically 0.2). Then you have to transform that into probabilities and minimize the cross-entropy to yield the 2D layout.

So even though the article says that you don't want a fixed radius on the high-dimensional balls but can use knns to "adapt to the local Riemann geometry", in practice software implementations do have an *additional* maximal radius (in practice, minimal simliarity), because otherwise you *always* end up connecting vertices/cells that are really different by "warping too much", and your topology goes to pot by basically folding the back of the sheet back onto itself. I had some fun with that cutoff in the past and biologically you can see that cells with correlation >0.9 should basically be merged (a la Louvain/Leiden), cells with 0.75-0.9 are "birds of a feather" (similar but not redundant), and cells <0.75 are already quite different from one another and are only connected together mainly to keep some appearance of "global geometry" on the space. So in my mind the whole "Riemaniann warping" technique works a little differently from how the mathematician imagined and would be an interesting topic of research in its own right. But I'm digressing!

Happy to chat on whatsapp/zoom if you prefer, but it sounds like you grasp this better than me already, unsurprisingly.

Reply:
The python umap implementation has an enormous amount of parameters you can tweak. One of the parameters to tweak for our umap could be the cutoff for the distance. One problem here is that I first want to have a basic umap working, before adding lots of parameters, but without all those parameters it just doesn't do what you expect. For now I'll just use 0.2 as a hardcoded cutoff.

umap just seems crazily general. Make fuzzy balls, use it to make a graph where close things are connected, and then shake your layout about until some function of the graph and layout is minimized.

Reply:
If I don't have the original data, any point in the umap graph can only be connected to other points if they're already connected in the knn graph, right? Because I have no clue about other distances.

Reply:
I think so

Reply:
>  The 0.2 is done before our part, so that's ok.

Yes that's part of my argument 😊

> umap just seems crazily general. Make fuzzy balls, use it to make a graph where close things are connected, and then shake your layout about until some function of the graph and layout is minimized.

See why I want to have it in igraph? I think it's just the first instance of a family of similar layouts (e.g. denseMAP, which I think is quite useless)